modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C

```diff
@@ -0,0 +1,95 @@
+
+InputParameters
+ReactorGeometryMeshBuilderBase::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addClassDescription("A base class that contains common members and methods for Reactor "
+                             "Geometry Mesh Builder mesh generators.");
+
+  return params;
+}
+
+ReactorGeometryMeshBuilderBase::ReactorGeometryMeshBuilderBase(const InputParameters & parameters)
+  : MeshGenerator(parameters)
+{
+}
+
+void
+ReactorGeometryMeshBuilderBase::initializeReactorMeshParams(const std::string reactor_param_name)
+{
+  bool first_call = _reactor_params != reactor_param_name;
+  if (first_call)
+    _reactor_params = reactor_param_name;
+
+  // Ensure that the user has supplied a valid ReactorMeshParams object
+  if (getMeshByName(_reactor_params) != nullptr)
+    mooseError("The reactor_params mesh is not of the correct type");
+
+  if (!hasMeshProperty("mesh_dimensions", _reactor_params) ||
+      !hasMeshProperty("mesh_geometry", _reactor_params))
+    mooseError("The reactor_params input must be a ReactorMeshParams type MeshGenerator\n Please "
+               "check that a valid definition and name of ReactorMeshParams has been provided.");
+
+  // Set parameters related to block name and block / region id mapping
+  _current_block_id = getReactorParam<subdomain_id_type>("current_block_id");
+  _name_id_map = getReactorParam<std::map<std::string, std::pair<subdomain_id_type, dof_id_type>>>(
+      "name_id_map");
+
+  // Set reactor_params_name metadata for use by future mesh generators
+  if (first_call)
+    declareMeshProperty("reactor_params_name", std::string(_reactor_params));
+}
+
+void
+ReactorGeometryMeshBuilderBase::initializeReactorMeshParams()
+{
+  initializeReactorMeshParams(_reactor_params);
+}
+
+void
+ReactorGeometryMeshBuilderBase::updateReactorMeshParams()
+{
+  this->declareMeshProperty(_reactor_params, "current_block_id", _current_block_id);
+  this->declareMeshProperty(_reactor_params, "name_id_map", _name_id_map);
+}
+
+dof_id_type
+ReactorGeometryMeshBuilderBase::getBlockId(const std::string block_name,
+                                           const dof_id_type region_id)
+{
+  if (_name_id_map.find(block_name) != _name_id_map.end())
+  {
+    // Block name exists in map, verify region_id matches one in map and return block_id
+    if (_name_id_map[block_name].second != region_id)
+      mooseError("Provided region id ",
+                 region_id,
+                 " for block name ",
+                 block_name,
+                 " does not match the one that was provided in an earlier mapping - ",
+                 _name_id_map[block_name].second);
+    return _name_id_map[block_name].first;
+  }
+  else
+  {
+    // Block name does not exist, return a new block id and store in _name_id_map
+    const auto new_block_id = _current_block_id++;
+    _name_id_map[block_name] = std::make_pair(new_block_id, region_id);
+    return new_block_id;
+  }
```

Comment:
@roystgnr 
they are adding this manual tracking of blocks within this MG.
Is this possible in libmesh already? @shikhar413 and I looked but didnt find. 

If not then I think it belongs there. What do you think?

Reply:
I'm not sure what the code here is trying to do.  LibMesh gives you a 1-to-1 mapping between subdomain names and subdomain ids (`MeshBase::subdomain_name()` and the how-did-I-let-Cody-slip-this-API-name-past `MeshBase::get_id_by_name()`), but we have no concept of "regions".

Reply:
Ok I’ll have another look. Thanks for pitching in

Reply:
For more context, these functions basically check that the subdomain name is mapped uniquely to (subdomain_id, region_id), where region_id is an extra element integer. For simplicity, assume that both subdomain name and region id is provided by the user. However, we would need to generate a _unique_ new subdomain id if it doens't already exist in the mesh. Is there a libmesh function that will allow us to do this? In the interim, I am manually creating a workaround for this, where `_current_block_id` will increment by 1 if a new block id is requested, and then this block name - (block_id, region_id) mapping is stored in `_name_id_map`

Reply:
`MeshBase::get_mesh_subdomains()` gives you an ordered set of subdomain ids, and you can just pick the max of that plus 1 to get a new unique subdomain id, but the set is only updated on `prepare_for_use()` (or if you want to do it manually, just `cache_elem_data()` is enough) since it requires a loop over elements and a parallel `set_union()`.

Reply:
so if we use what Roy suggests, what does this map provide? A way to check that region ids are set consistently across block IDs ?

I think this map might actually be counter productive, since it forces a unique mapping block name -> region id and the whole purpose of region id is not have this uniqueness. We want multiple region ids in a given block, because we can only have so many blocks before we stop scaling

Reply:
@GiudGiud , there seems to be some confusion about the purpose of _name_id_map. The map is there just to check that the user has not tried to define a block_name - region_id pairing that is inconsistent with previously defined pairs in the other RGMB MG's. For this reason, we need this variable to persist throughout the duration of the RGMB mesh generation stage (from pin mesh generation all the way up to core mesh generation). However, notice that the key in the map is the block name, not the region id. Therefore, you can have multiple blocks that share the same region id. In fact, this is done in the code (and you can see this in some of the unit test cases as well). For triangular regions, the "_TRI" suffix is appended to the block name, but they will share the same region ID as the quad region, so you can have the following block name - region id pairings:

RGMB_BLOCK_1 : 1
RGMB_BLOCK_TRI : 1

In both cases, we need to make sure the two block names are associated with unique block ids, hence the discussion with Roy about how to generate block ids that do not cause any contention with previously defined MGs. The user should not have to worry about what these block id's are, they are used and tracked internally by the RGMB MG's and stored in the map as a block_name: (block_id, region_id) mapping.

One other thing to note is that all of these mappings should only be relevant for RGMB MG's, i.e. none of the other MG's need to have this design constraint enforced. These MG's were designed under the assumption that these MGs would be completely self-contained so that other MG's would not need to be used before, during, or after the RGMB MG stage, but if users choose to do so, then it is up to them that the other MGs are consistent with how RGMB has generated its mesh

Reply:
> notice that the key in the map is the block name, not the region id

that is the problem. That means you can only have as many region ids as you have blocks. 
The key, if you wanted to do this, which we're more and more leaning against should be the region id because that is the one that you have plenty of.

Reply:
I'm not sure I understand. How would you represent a block with a mixture of tri and quad elements? If the key were a unique region id among the two, then there'd be no way to represent both elements as having a shared region id since the unique key would have to point to either the quad elements or the tri elements.

Maybe there is a misunderstanding in terms of what region id represents? Here, region id is a single extra element id to represent like materials, like in Griffin. For example, if we have a fuel pin with tri and quad elements, these would have to be unique blocks, but they can be assigned the same region ID to indicate that these elements have the same material. All other extra elements integers (such as pin id, assembly id, plane id etc) are assigned on their own, independent of the region id, and therefore can be combined together for a single block.

> That means you can only have as many region ids as you have blocks.

I don't really see how the number of region ids could outnumber the number of blocks

Reply:
The whole purpose of using region ids is to be able to be able to have multiple region ids within a block.
For example in an assembly you could have 1 block: fuel and 284 different region ids, which correspond to the cross section regions. Your second paragraph shows we agree on this purpose. This map precludes this, because you can only have less regions than blocks when they map to blocks with a block_name key. 

You do not represent a block with a mixture of tri and quad, you use two different blocks for this but then you use the same region id if appropriate. This is the exception case where you need one more block than region id. The general case is that we want more region ids than blocks. 

You can have a key of region id and a vector of blocks as the item. That is again if we assume we want this map.

Maybe we should meet. Available tomorrow after (Wednesday) or Thursday 2pm +.

Reply:
I think I see what you're saying. I have to think about this more but with this new design, the need for this map isn't obvious to me anymore, since we no longer are imposing any constraints on blocks / region ids. The only additional functionality that I think we should keep is just having the ability to generate unique non-clashing block ids for every block name that isn't already in the mesh, which I think can be done entirely on the libMesh side without the use of this map.

Down for a call, tomorrow afternoon works for me

Reply:
@roystgnr is there a way to call `cache_elem_data()` in the mesh generator constructor in order to figure out what the max block id so far is? These subdomain id values are set in the constructors of the mesh generators in this PR by calling `addMeshSubgenerator` for existing mesh generators, so this max value is needed before the mesh is actually generated in order to ensure that these subdomain ids don't potentially conflict with other MGs later on

Reply:
I'm playing around with the `cache_elem_data()` and  `get_mesh_subdomains()` functions, it looks like it only returns the subdomain ids of mesh created within the scope of the current mesh generator block, is there a way to return subdomain ids of _all mesh generators up to the current mesh generator_ under consideration? Basically if there are two pin mesh generators that are called successively (and not necessarily in communication with each other), I would like to know the block ids used by the first pin mesh generator so that the second pin mesh generator can set its block ids without conflicting with those set by the first mesh generator.

Reply:
libMesh doesn't know what a mesh generator even is.  `cache_elem_data()` will calculate on whichever mesh you call it on, `get_mesh_subdomains()` will return data for whichever mesh you call it on; if you have multiple `MeshBase` objects being operated on then you'll need to make those calls on each (and merge the results of the get afterwards).  If your first pin mesh generator is an input to your second pin mesh generator then you've got an easy time of it; if not you're out of luck.

Your best bet in the latter case might be to handle any conflicts at the point of merging.  This is what libMesh does when stitching two meshes together, for instance; if one mesh has an "element 1" and the another has a different "element 1" then we don't just throw our hands up and say the meshes conflict, we renumber the latter's elements (calculating an offset to add to element ids, node ids, unique_ids) to solve the conflict.

Reply:
Thanks for that explanation @roystgnr. That makes a lot of sense, unfortunately for us, the situation you described in the latter case is what is ultimately required for our purposes so some additional effort is definitely required.

@GiudGiud do you have any thoughts on this? Right now I am thinking that in the constructor we assign block id's as a temporary value (something like UINT_MAX - 1 - elem_index) in the constructor, and then try to merge all block name / id info and remap all element ids/names in the `generate` function later. The implementation won't be very straightforward or clean, however. It'd definitely be easier to let the user try to set the block id's with the block names and region ids instead. Also, given that a mesh generator has no idea what block id's have been assigned up to the point when it's been called, this also does not address the fundamental issue of, "how do we assign block ids to RGMB MGs without conflict if other MGs are going to be called before, during, or after the RGMB generation stage?"

Reply:
So a few things:
- I m not sure you actually want different block IDs by default. We can only handle so many blocks, so the important part is to get different region IDs by default.
- Since you are using the subgenerator system, cant the block IDs be set at the time you are adding these?
- merging meshes and making sure blocks are different is not necessarily "unclean". To me it is less object oriented and more unclean to be generating a pin mesh while keeping track of the mesh (block ids) for an entire reactor.

Reply:
So I don't think setting block IDs at the sub generator level will not be all that straightforward, since the IDs and names would need to be set together and there is no way to ensure that there isn't a conflict with later MGs. To me, it seems like temporary IDs and names will have to be set at the sub generator level and then reset at the generate level.

Reply:
ok let's go with this then.

Reply:
> Your best bet in the latter case might be to handle any conflicts at the point of merging. This is what libMesh does when stitching two meshes together, for instance; if one mesh has an "element 1" and the another has a different "element 1" then we don't just throw our hands up and say the meshes conflict, we renumber the latter's elements (calculating an offset to add to element ids, node ids, unique_ids) to solve the conflict.

@roystgnr can you provide more information on this comment? Does it work for mesh stitching where two MG's share the same block ID but have different names? As I implement the features requested by @GiudGiud, I am running into the issue where pin1 has let's say block ID's numbers from 10001 to 10005 and pin2 also has block IDs numbered from 10001 to 10005 but the blocks all have unique block names across the two pins. 

However, when these pins are stitched into an assembly, instead of renumbering the block IDs from 10001 to 10010 and preserving these unique block names, the block IDs are merged so that block names for pin2 are removed and the two pins continue to share the same block IDs from 10001 to 10005. I would need this functionality of remapping block IDs instead of merging them, since pin2 has no information about pin1 anymore, so an appropriate block ID offset for pin 2 cannot be known a priori. Could you please advise on if this can be done?

Reply:
in the merge/rename routine cant you detect that they have a different block name and then nor merge the block IDs.

Or are you saying that the stitching is problematic because it thinks blocks with the same block IDs but with a different name should be merged?
In the latter case we d need to patch libmesh I think

Reply:
So if you have a mesh block like this:

```
[Mesh]
  [pin1]
    type = PolygonConcentricCircleMeshGenerator
    polygon_size = 1.42063
    num_sides = 4
    ring_radii = '0.2 0.4'
    ring_block_ids = '10000 10001'
    ring_intervals = '1 1'
    ring_block_names = 'pin1_r0 pin1_r1'
    background_block_ids = 100002
    background_intervals = 1
    background_block_names = 'pin1_r2'
  []

  [pin2]
    type = PolygonConcentricCircleMeshGenerator
    polygon_size = 1.42063
    num_sides = 4
    ring_radii = '0.4 0.6'
    ring_block_ids = '10000 10001'
    ring_intervals = '1 1'
    ring_block_names = 'pin2_r0 pin2_r1'
    background_block_ids = 100002
    background_intervals = 1
    background_block_names = 'pin2_r2'
  []

  [assembly]
    type = CartesianIDPatternedMeshGenerator
    id_name = pin_id
    assign_type = cell
    inputs = 'pin1 pin2'
    pattern = '0 0;
               0 1;'
  []
[]
```

I was hoping to use the block name to figure out which radial level and pin ID I am currently iterating over in the merge/rename routine of `AssemblyMeshGenerator` so as to set the block names and region IDs according to what the user wants, but the block names for id's 10000 to 10002 get merged into one and so when you iterate over the pin elements in `AssemblyMeshGenerator::generate()`, at this point it's unclear whether id 10000 refers to pin1 or pin2. This block id merging leads to more complications once we start to incorporate background and duct regions in multiple assemblies as well. One possible workaround would be to tag the radial index of each pin and assembly element as an extra element integer (0+offset for radial indices in the pin, 1000+offset for assembly background and duct elements), and use this radial id, pin id, and axial_id to set the block names and region ids. What do you think?

Reply:
> Does it work for mesh stitching where two MG's share the same block ID but have different names?

The feature I described doesn't do anything at all to subdomain ids.  If we stitch two meshes and both have a subdomain 3, we assume that the final mesh should also have a subdomain 3 which is the union of the inputs, because subdomain ids don't confict in general.  I'm pointing out what we do for the values (id() and unique_id()) which do conflict, and suggesting that since in your case subdomain ids conflict too you might try an analogous tactic.

Frankly, in the case of the input file you posted, I think the correct behavior would be to die with an error.  We do require that subdomain ids and subdomain names are in 1-to-1 (or 1-to-0 when no name exists) correspondence, so we can't have two names for the same id, but the first input mesh clearly wants one name for that id and the second mesh clearly wants another.  If the input file had created automatically-chosen numeric ids rather than specifying exact numerical ids then I'd say the tactic I recommend above would apply: just choose a new number to go with the name of any subdomain id that conflicts, and renumber ... but the input file does specify exact ids and so it's not fair to the user to change them behind the user's back.